### PR TITLE
Raise an event when redirect trigger detected (#879)

### PR DESCRIPTION
**Pull request by AlexanderMoskovkin**

### CODE REVIEW

File: src/client/redirect-watch.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import EventEmiter from './utils/event-emitter';
+import { parseProxyUrl } from '../utils/url';
+
+const HASH_RE = /#.*$/;
+
+export default class RedirectWatch extends EventEmiter {
```
Comment (inikulin) [inline src/client/redirect-watch.js]: It's better be called PageNavigationWatch. Redirect is a different thing

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot /retest

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
The PR based on commits for https://github.com/DevExpress/testcafe-hammerhead/issues/667 (that were created by @churkin, but were reverted in https://github.com/DevExpress/testcafe-hammerhead/commit/3b536c72c765970b609c2b94df0325cb3b4442b2).

I've moved base things from there, but simplified things in situation when a handler prevents an event.

/cc @churkin @miherlosev @VasilyStrelyaev

File: src/client/page-navigation-watch.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        var target       = nativeMethods.getAttribute.call(el, domProcessor.getStoredAttrName('target')) ||
+                           nativeMethods.getAttribute.call(el, 'target');
+
+        if (target) {
+            if (!ElementSandbox._isKeywordTarget(target))
+                targetWindow = windowsStorage.findByName(target) || window;
+            else if (target === '_top' || target === '_blank')
+                targetWindow = window.top;
+            else if (target === '_parent')
+                targetWindow = window.parent;
```
Comment (churkin) [inline src/client/page-navigation-watch.js]: What if the `window.parent` is null?

File: src/client/page-navigation-watch.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+                    // NOTE: the event keep the defaultPrevented flag between event handlers in all
+                    // browsers except IE. In IE it is reset to false before the next handler execution.
+                    if (!e.defaultPrevented && !prevented)
+                        onFormSubmit(e.target);
+                });
+        });
+    }
+
+    static _getTargetWindow (el) {
```
Comment (churkin) [inline src/client/page-navigation-watch.js]: I think it is util funciton

File: src/client/page-navigation-watch.js (hunk 2 above)
Comment (AlexanderMoskovkin) [inline src/client/page-navigation-watch.js]: If a window does not have a parent, its parent property is a reference to itself (
[proof](https://developer.mozilla.org/en/docs/Web/API/Window/parent)). Do you know a case when `window.parent === null` ?

Comment (churkin) [inline src/client/page-navigation-watch.js]: For the top window

Comment (AlexanderMoskovkin) [inline src/client/page-navigation-watch.js]: Writing again: If a window does not have a parent, its parent property is a reference to itself =)

In chrome, ff, ie for the top window `window.parent === window`. In what browser or case `window.parent === null`?

File: src/client/page-navigation-watch.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            };
+
+            eventSandbox.on(eventSandbox.EVENT_PREVENTED_EVENT, onPreventDefault);
+
+            nextTick()
+                .then(() => {
+                    eventSandbox.off(eventSandbox.EVENT_PREVENTED_EVENT, onPreventDefault);
+
+                    // NOTE: the event keep the defaultPrevented flag between event handlers in all
+                    // browsers except IE. In IE it is reset to false before the next handler execution.
```
Comment (VasilyStrelyaev) [inline src/client/page-navigation-watch.js]:
```
                    // NOTE: the defaultPrevented flag is saved between event raises in all
                    // browsers except IE. In IE, it is reset to false before the next handler is executed.
```

File: src/client/page-navigation-watch.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                };
+
+                eventSandbox.on(eventSandbox.EVENT_PREVENTED_EVENT, onPreventDefault);
+
+                nextTick()
+                    .then(() => {
+                        eventSandbox.off(eventSandbox.EVENT_PREVENTED_EVENT, onPreventDefault);
+
+                        // NOTE: the event keep the defaultPrevented flag between event handlers in all
+                        // browsers except IE. In IE it is reset to false before the next handler execution.
```
Comment (VasilyStrelyaev) [inline src/client/page-navigation-watch.js]: see above

File: test/client/fixtures/page-navigation-watch-test.js [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+            start();
+        })
+        .catch(function () {
+            ok(true);
+            start();
+        });
+});
+
+if (!browserUtils.isIE || browserUtils.isMSEdge) {
+    // NOTE: we don't catch event preventing via "return false" in inline handlers in IE yet
```
Comment (VasilyStrelyaev) [inline test/client/fixtures/page-navigation-watch-test.js]:
```
    // NOTE: we can't detect if default handling was cancelled by returning 'false' from an inline handler in IE yet
```

File: test/client/fixtures/sandbox/node/attributes-test.js [hunk 7]
```diff
@@ -23,6 +23,47 @@ QUnit.testDone(function () {
     iframeSandbox.off(iframeSandbox.RUN_TASK_SCRIPT, initIframeTestHandler);
 });
 
+// NOTE: IE11 has a strange bug that does not allow test
```
Comment (VasilyStrelyaev) [inline test/client/fixtures/sandbox/node/attributes-test.js]:
```
// NOTE: IE11 has a strange bug that does not allow this test to pass
```

**Comment by AlexanderMoskovkin** [comment]
FPR /cc @churkin @miherlosev @inikulin @VasilyStrelyaev

**Pull request merged**